provider/recent_hash_store.go

```diff
@@ -0,0 +1,98 @@
+package provider
+
+import (
+	"context"
+	"github.com/bloXroute-Labs/solana-trader-client-go/connections"
+	pb "github.com/bloXroute-Labs/solana-trader-client-go/proto"
+	log "github.com/sirupsen/logrus"
+	"sync"
+	"time"
+)
+
+type blockHashProvider func(ctx context.Context) (*pb.GetRecentBlockHashResponse, error)
+type blockHashStreamProvider func(ctx context.Context) (connections.Streamer[*pb.GetRecentBlockHashResponse], error)
+
+type recentBlockHashStore struct {
+	mutex              sync.RWMutex
+	hashProvider       blockHashProvider
+	hashStreamProvider blockHashStreamProvider
+	hash               string
+	hashTime           time.Time
+	hashExpiryDuration time.Duration
+}
+
+func newRecentBlockHashStore(
+	hashProvider blockHashProvider,
+	streamProvider blockHashStreamProvider,
+	opts RPCOpts,
+) *recentBlockHashStore {
+	return &recentBlockHashStore{
+		mutex:              sync.RWMutex{},
+		hashProvider:       hashProvider,
+		hashStreamProvider: streamProvider,
+		hash:               "",
+		hashTime:           time.Time{},
+		hashExpiryDuration: opts.CacheBlockHash,
+	}
+}
+
+func (s *recentBlockHashStore) run() {
+	ctx := context.Background()
```

Comment:
should probably pass this context through the store when initializing it, as there's currently no way to cancel this loop

Reply:
I'm ok with requiring a `ctx` object for each provider type

Reply:
yeah, in general it's worth even to pass context when creating client, that's not a part of the api